### PR TITLE
Faster Gif Decoding

### PR DESCRIPTION
**Pull request by JimBobSquarePants**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
Nothing fancy just a bit of a performance improvement reading and writing gifs. 33% faster decoding but no noticeable change with the encoder. 

We've got 250 GC Generation 2 collects per 1k Operations when encoding that I haven't identified the reason for yet. Possibly caused by multiple short writes to the stream.

**Before**

Method | Runtime |     TestImage |       Mean |      Error |   StdDev | Scaled | ScaledSD |   Gen 0 | Allocated |
--------------------- |-------- |-------------- |-----------:|-----------:|---------:|-------:|---------:|--------:|----------:|
'System.Drawing Gif' |    Core | Gif/rings.gif |   440.1 us |   966.4 us | 54.60 us |   1.00 |     0.00 | 51.2695 |  105.3 KB |
'ImageSharp Gif' |    Core | Gif/rings.gif | 1,231.9 us |   594.7 us | 33.60 us |   2.83 |     0.27 |       - |   2.04 KB |
|         |               |            |            |          |        |          |         |           |
'System.Drawing Gif' |     Clr | Gif/rings.gif |   386.2 us |   198.9 us | 11.24 us |   1.00 |     0.00 | 51.2695 | 105.44 KB |
'ImageSharp Gif' |     Clr | Gif/rings.gif | 1,317.2 us | 1,458.3 us | 82.40 us |   3.41 |     0.19 |       - |   2.09 KB |

**After** 

Method | Runtime |     TestImage |     Mean |    Error |   StdDev | Scaled |   Gen 0 | Allocated |
--------------------- |-------- |-------------- |---------:|---------:|---------:|-------:|--------:|----------:|
'System.Drawing Gif' |    Core | Gif/rings.gif | 391.2 us | 32.14 us | 1.816 us |   1.00 | 51.2695 |  105.3 KB |
'ImageSharp Gif' |    Core | Gif/rings.gif | 909.6 us | 83.44 us | 4.714 us |   2.32 |       - |   2.04 KB |
|         |               |          |          |          |        |         |           |
'System.Drawing Gif' |     Clr | Gif/rings.gif | 392.9 us | 25.92 us | 1.464 us |   1.00 | 51.2695 | 105.44 KB |
'ImageSharp Gif' |     Clr | Gif/rings.gif | 935.4 us | 83.61 us | 4.724 us |   2.38 |  0.9766 |   2.09 KB |

### CODE REVIEW

File: src/ImageSharp/Formats/Gif/GifEncoderCore.cs [hunk 1]
```diff
@@ -132,17 +134,19 @@ public void Encode<TPixel>(Image<TPixel> image, Stream stream)
         /// <returns>
         /// The <see cref="int"/>.
         /// </returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (antonfirsov) [inline src/ImageSharp/Formats/Gif/GifEncoderCore.cs]: I know, I did it several times, but it's most likely not worth to inline methods with complex loops inside them.

File: src/ImageSharp/Formats/Gif/GifEncoderCore.cs [hunk 2]
```diff
@@ -336,15 +341,19 @@ private void WriteColorTable<TPixel>(QuantizedFrame<TPixel> image, EndianBinaryW
             var rgb = default(Rgb24);
             using (IManagedByteBuffer colorTable = this.memoryManager.AllocateManagedByteBuffer(colorTableLength))
             {
-                Span<byte> colorTableSpan = colorTable.Span;
+                // TODO: Pixel operations?
+                ref TPixel paletteRef = ref MemoryMarshal.GetReference(image.Palette.AsSpan());
+                ref byte colorTableRef = ref MemoryMarshal.GetReference(colorTable.Span);
```
Comment (antonfirsov) [inline src/ImageSharp/Formats/Gif/GifEncoderCore.cs]: By casting the `byte` reference to `Rgb24` reference, it's possible to have just one `Unsafe.Add()` inside the loop, also eliminating `i * 3`.

Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Gif/GifEncoderCore.cs]: Damn that's nice!

**Comment by iamcarbon** [comment]
@JimBobSquarePants -- I had actually started work to eliminate the EndianBinaryWriter entirely from GifEncoder. It's a heavy abstraction that we don't need here. Can / will take a stab in my next PR after this is merged.

File: src/ImageSharp/Formats/Gif/GifEncoderCore.cs [hunk 3]
```diff
@@ -336,15 +335,14 @@ private void WriteColorTable<TPixel>(QuantizedFrame<TPixel> image, EndianBinaryW
...
                 for (int i = 0; i < pixelCount; i++)
                 {
-                    int offset = i * 3;
-                    image.Palette[i].ToRgb24(ref rgb);
-                    colorTableSpan[offset] = rgb.R;
-                    colorTableSpan[offset + 1] = rgb.G;
-                    colorTableSpan[offset + 2] = rgb.B;
+                    ref TPixel entry = ref Unsafe.Add(ref paletteRef, i);
+                    entry.ToRgb24(ref rgb);
+                    Unsafe.Add(ref rgb24Ref, i);
```
Comment (antonfirsov) [inline src/ImageSharp/Formats/Gif/GifEncoderCore.cs]: You forgot to assign to assign the reference to a value.

Doesn't this break any tests ??? :fearful:
If not, I think we should add them ASAP. Unsafe optimizations could not be carried out safely without proper coverage.

Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Gif/GifEncoderCore.cs]: I think you might be reading this wrong. I’m getting the expected output

Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Gif/GifEncoderCore.cs]: Hold on, you’re right! This path is definitely called, how the hell did I produce correct output. 

I’ll update and investigate

File: tests/ImageSharp.Tests/Formats/Gif/GifEncoderTests.cs [hunk 4]
```diff
@@ -22,28 +24,43 @@ public void EncodeGeneratedPatterns<TPixel>(TestImageProvider<TPixel> provider)
...
+
+                // Always save as we need to compare the encoded output.
+                provider.Utility.SaveTestOutputFile(image, "gif", encoder);
+            }
+
+            // Compare encoded result
+            string path = provider.Utility.GetTestOutputFileName("gif", null, true);
+            using (var encoded = Image.Load(path))
+            {
+                encoded.CompareToReferenceOutput(ValidatorComparer, provider, null, "gif");
```
Comment (antonfirsov) [inline tests/ImageSharp.Tests/Formats/Gif/GifEncoderTests.cs]: Does this fail if you reintroduce the mistake in `WriteApplicationExtension()`?

If yes, I'm happy to merge this!

Comment (JimBobSquarePants) [reply tests/ImageSharp.Tests/Formats/Gif/GifEncoderTests.cs]: Yeah you end up with a black and white image

**Pull request merged**